src/main.rs

```diff
@@ -47,125 +96,265 @@ impl LogLevel {
-                // combine and print stderr/stdout
-                log::warn!("wanted script failed...");
-            }
-        }
+
+    if !path_exists {
+        bail!("required.d not found");
     }
-    // if a command with restart option in systemd fails to start we don't get it as "failed"
-    // reversing the check makes sure that if by the time After=multi-user the service isn't running then it's failing at least
-    let output = Command::new("systemctl")
-        .arg("list-units")
-        .arg("--state")
-        .arg("active")
-        .arg("--no-page")
-        .arg("--output")
-        .arg("json")
-        .output()?;
-    let services: Vec<ServiceStatus> = serde_json::from_str(&String::from_utf8(output.stdout)?)?;
-    let ss: Vec<String> = services.iter().map(|x| x.unit.clone()).collect();
-    let active_units: HashSet<String> = HashSet::from_iter(ss);
-    for service in ["sshd.service", "NetworkManager.service"] {
-        if !active_units.contains(service) {
-            log::warn!("service {} failed, see journal", service);
-            failure = true;
+
+    for path in GREENBOOT_INSTALL_PATHS {
+        let greenboot_wanted_path = format!("{path}/check/wanted.d/");
+        if let Some(e) = run_scripts("wanted", &greenboot_wanted_path).err() {
+            log::error!("wanted script runner error:{}", e.to_string());
         }
     }
-    if failure {
-        for path in ["/etc/greenboot/red.d/*.sh"] {
-            for entry in glob(path)?.flatten() {
-                log::info!("running red check {}", entry.to_string_lossy());
-                let output = Command::new("bash").arg("-C").arg(entry).output()?;
-                if !output.status.success() {
-                    // combine and print stderr/stdout
-                    log::warn!("red script failed...");
-                }
-            }
-        }
-        log::warn!("SYSTEM is UNHEALTHY. Rolling back and rebooting...");
-        Command::new("rpm-ostree").arg("rollback").status()?;
-        reboot()?;
-        return Ok(());
-    }
-    for path in ["/etc/greenboot/green.d/*.sh"] {
-        for entry in glob(path)?.flatten() {
-            log::info!("running green check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
-            if !output.status.success() {
-                // combine and print stderr/stdout
-                log::warn!("green script failed...");
-            }
-        }
+
+    if script_failure {
+        bail!("health-check failed!");
     }
     Ok(())
 }
 
-fn reboot() -> Result<(), Error> {
-    Command::new("systemctl").arg("reboot").spawn()?;
+/// runs the scripts in red.d when health-check fails
+fn run_red() -> Result<()> {
+    for path in GREENBOOT_INSTALL_PATHS {
+        let red_path = format!("{path}/red.d/");
+        if let Some(e) = run_scripts("red", &red_path).err() {
+            log::error!("red script runner error: {}", e.to_string());
```

Comment:
In this function you are never returning `Err` since you are not propagating it

Reply:
returned error.